packages/versioning/src/versioning.ts

```diff
@@ -78,18 +89,74 @@ export function $madeOptional(context: DecoratorContext, t: Type, v: EnumMember)
   program.stateMap(madeOptionalKey).set(t, version);
 }
 
+function toVersion(p: Program, t: Type, v: ObjectType): Version | undefined {
+  const [namespace] = getVersions(p, t);
+  if (namespace === undefined) {
+    return undefined;
+  }
+  return getVersionForNamespace(
+    p,
+    v,
+    (namespace.projectionBase as Namespace) ?? namespace
+  ) as Version;
+}
+
+function getRenamedFrom(p: Program, t: Type): Array<RenamedFrom> | undefined {
+  return p.stateMap(renamedFromKey).get(t) as Array<RenamedFrom>;
+}
+
 /**
  * @returns version when the given type was added if applicable.
  */
 export function getRenamedFromVersion(p: Program, t: Type): Version | undefined {
-  return p.stateMap(renamedFromKey).get(t)?.v;
+  p.reportDiagnostic({
+    code: "deprecated",
+    message: "Deprecated: getRenamedFromVersion is deprecated. Use getRenamedFromVersions instead.",
+    severity: "error",
+    target: t,
+  });
+  return p.stateMap(renamedFromKey).get(t)?.[0].version;
+}
+
+/**
+ * @returns the list of versions for which this decorator has been applied
+ */
+export function getRenamedFromVersions(p: Program, t: Type): Version[] | undefined {
+  return getRenamedFrom(p, t)?.map((x) => x.version);
 }
 
 /**
  * @returns get old renamed name if applicable.
  */
-export function getRenamedFromOldName(p: Program, t: Type): string {
-  return p.stateMap(renamedFromKey).get(t)?.oldName ?? "";
+export function getRenamedFromOldName(p: Program, t: Type, v: ObjectType): string {
+  p.reportDiagnostic({
+    code: "deprecated",
+    message: "Deprecated: getRenamedFromOldName is deprecated. Use getNameAtVersion instead.",
+    severity: "error",
+    target: t,
+  });
+  return p.stateMap(renamedFromKey).get(t)?.[0].oldName ?? "";
+}
+
+/**
+ * @returns get old name if applicable.
+ */
+export function getNameAtVersion(p: Program, t: Type, v: ObjectType): string {
+  const version = toVersion(p, t, v);
+  const allValues = getRenamedFrom(p, t);
+  if (!allValues || !version) return "";
+
+  let oldName = "";
+  const targetIndex = version.index;
+
+  for (const val of allValues) {
+    const index = val.version.index;
+    if (targetIndex < index) {
+      oldName = val.oldName;
+      break;
```

Comment:
```suggestion
      return val.oldName;
```

isn't that the same? that way get rid of `oldName`

Reply:
Updated